File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-rec-iter-fromgeneral.rb
# LeetCode #94 - Binary Tree Inorder Traversal
# https://leetcode.com/problems/binary-tree-inorder-traversal/
# Special case of a general iteratively implemented recursive solution.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
  out = []
  traverse(root, ->(_) { }, ->(val) { out << val }, ->(_) { })
  out
end

# A stack frame for iteratively implemented recursive binary tree traversal.
class Frame
  attr_reader :node
  attr_accessor :state

  def initialize(node, state)
    @node = node
    @state = state
  end
end

def traverse(root, f_pre, f_in, f_post)
  stack = []
  stack << Frame.new(root, :go_left) if root

  until stack.empty?
    frame = stack.last

    case frame.state
    when :go_left
      f_pre.call(frame.node.val)
      stack << Frame.new(frame.node.left, :go_left) if frame.node.left
      frame.state = :go_right
    when :go_right
      f_in.call(frame.node.val)
      stack << Frame.new(frame.node.right, :go_left) if frame.node.right
      frame.state = :retreat
    when :retreat
      f_post.call(frame.node.val)
      stack.pop
    end
  end
end

File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-rec-iter-fromgeneral.java
// LeetCode #94 - Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/
// From general iteratively implemented recursive DFS (state machine).

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        dfs(root, Solution::noop, values::add, Solution::noop);
        return values;
    }

    private static void noop(int _value) {
        // Do nothing.
    }

    private static enum State {
        GO_LEFT,
        GO_RIGHT,
        RETREAT
    }

    private static final class Frame {
        Frame(TreeNode node) {
            this.node = node;
            this.state = State.GO_LEFT;
        }

        final TreeNode node;
        State state;
    }

    private static void dfs(TreeNode root,
                            IntConsumer preorderAction,
                            IntConsumer inorderAction,
                            IntConsumer postorderAction) {
        if (root == null) return;

        var stack = Collections.asLifoQueue(new ArrayDeque<Frame>());
        stack.add(new Frame(root));

        while (!stack.isEmpty()) {
            var frame = stack.element();

            switch (frame.state) {
            case GO_LEFT:
                preorderAction.accept(frame.node.val);
                frame.state = State.GO_RIGHT;
                if (frame.node.left != null)
                    stack.add(new Frame(frame.node.left));
                break;

            case GO_RIGHT:
                inorderAction.accept(frame.node.val);
                frame.state = State.RETREAT;
                if (frame.node.right != null)
                    stack.add(new Frame(frame.node.right));
                break;

            case RETREAT:
                postorderAction.accept(frame.node.val);
                stack.remove();
                break;
            }
        }
    }
}
